src/compiler/checker.ts

```diff
@@ -19103,9 +19103,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const mapper = createTypeMapper([getTypeParameterFromMappedType(objectType)], [index]);
         const templateMapper = combineTypeMappers(objectType.mapper, mapper);
         const instantiatedTemplateType = instantiateType(getTemplateTypeFromMappedType(objectType.target as MappedType || objectType), templateMapper);
-        const isOptional = getMappedTypeOptionality(objectType) > 0 || (isGenericType(objectType) ?

```

Comment:
In the given repro the immediate modifiers type has `-?` for which `getMappedTypeOptionality` returns `-1`. So the first condition was not satisfied and the compiler proceeded to compute `getCombinedMappedTypeOptionality`. This time it was called with the modifiers type and not the original object type. Since that was a mapped type too it got its modifiers type etc.

Essentially this was skipping the outermost mapped type modifiers in this scenario - the ones that were supposed strip the optionality.